test/distributed/fsdp/test_fsdp_backward_prefetch.py

```diff
@@ -0,0 +1,126 @@
+# Owner(s): ["oncall: distributed"]
+
+import sys
+from unittest.mock import patch
+
+import torch
+import torch.nn as nn
+from torch import distributed as dist
+from torch.distributed.fsdp import BackwardPrefetch, FullyShardedDataParallel as FSDP
+from torch.distributed.fsdp._runtime_utils import (
+    _get_handle_to_prefetch,
+    _get_training_state,
+)
+from torch.distributed.fsdp.flat_param import HandleTrainingState
+from torch.distributed.fsdp.wrap import ModuleWrapPolicy
+from torch.testing._internal.common_distributed import skip_if_lt_x_gpu
+from torch.testing._internal.common_fsdp import FSDPTest
+from torch.testing._internal.common_utils import run_tests, TEST_WITH_DEV_DBG_ASAN
+
+NUM_ITERS = 1
+
+if not dist.is_available():
+    print("Distributed not available, skipping tests", file=sys.stderr)
+    sys.exit(0)
+
+if TEST_WITH_DEV_DBG_ASAN:
+    print(
+        "Skip dev-asan as torch + multiprocessing spawn have known issues",
+        file=sys.stderr,
+    )
+    sys.exit(0)
+
+
+class TestBackwardPrefetch(FSDPTest):
+    @property
+    def world_size(self):
+        return 2
+
+    def _dist_train(self, backward_prefetch=BackwardPrefetch.BACKWARD_PRE):
+        rank = self.rank
+        orig_get_handle_to_prefetch = _get_handle_to_prefetch
+
+        torch.manual_seed(0)
+        policy = ModuleWrapPolicy(
+            {nn.TransformerEncoderLayer, nn.TransformerDecoderLayer}
+        )
+        model = FSDP(
+            nn.Transformer(d_model=1024, nhead=8, device="cuda"),
+            device_id=torch.cuda.current_device(),
+            auto_wrap_policy=policy,
+            use_orig_params=True,
+            backward_prefetch=backward_prefetch,
+        )
+        optim = torch.optim.SGD(model.parameters(), lr=1e-2)
+
+        # prepare input
+        torch.manual_seed(rank + 1)
+        src = torch.randn((10, 1, 1024), device="cuda")
+        tgt = torch.randn((20, 1, 1024), device="cuda")
+
+        # monkey patch
+        non_none_handle_count = 0
+        func_call_count = 0
+
+        def patched_get_handle_to_prefetch(*args, **kwargs):
+            handle = orig_get_handle_to_prefetch(*args, **kwargs)
```

Comment:
Sorry that I am only thinking about how to do this now.

I think it would be valuable to hard code that the sequence of prefetched handles is the expected one. This test only checks that _some_ handles were prefetched, not that they were reasonable handles.

Let me know if you need help figuring out how map modules to handles.

Reply:
To verify the sequence of prefetched handles, I added DECODER_PARAM_FQNS and ENCODER_PARAM_FQNS